astropy/coordinates/baseframe.py

```diff
@@ -1227,14 +1209,102 @@ def transform_to(self, new_frame):
+            A new object with the coordinate data represented in the
+            ``new_frame`` system.
+
+        Raises
+        ------
+        ValueError
+            If there is no possible transformation route.
+        """
+        if self._data is None:
+            raise ValueError('Cannot transform a frame with no data')
+
+        new_frame, trans = self._conform_and_get_transformation(new_frame)
+
+        if trans is None:
+            # no special transform needed, but should update frame info
+            return new_frame.realize_frame(self.data)
+
         return trans(self, new_frame)
 
+    def _get_equivalent_affine_params_to(self, new_frame, *further_frames):
+        """
+        Return equivalent affine parameters for transforming to a new frame.
+
+        The parameters for an affine transformation are a 3 x 3 Cartesian
+        transformation matrix and a displacement vector, which can include an
+        attached velocity.  Either type of parameter can be ``None``.
+
+        Every transformation step to the new frame must be an affine
+        transformation.  Further frames can be provided as additional arguments,
+        and the returned parameters will be for the full transformation sequence
+        of the arguments in the provided order.
+        """
+        from .errors import ConvertError
+
+        new_frame, trans = self._conform_and_get_transformation(new_frame)
+
+        affine_params = (None, None)
+
+        if trans is not None:
+            curr_frame = self
+
+            for t in trans.transforms:
+                if not isinstance(t, BaseAffineTransform):
+                    raise ConvertError(f"The transformation "
+                                       f"from {t.fromsys.__name__} "
+                                       f"to {t.tosys.__name__} "
+                                       f"is a {t.__class__.__name__} "
+                                       "rather than an affine transformation.")
+
+                # build an intermediate frame with attributes taken from either
+                # self, or if not there, `new_frame`, or if not there, use
+                # the defaults
+                frattrs = {}
+                for inter_frame_attr_nm in t.tosys.get_frame_attr_names():
+                    if hasattr(new_frame, inter_frame_attr_nm):
+                        attr = getattr(new_frame, inter_frame_attr_nm)
+                        frattrs[inter_frame_attr_nm] = attr
+                    elif hasattr(self, inter_frame_attr_nm):
+                        attr = getattr(self, inter_frame_attr_nm)
+                        frattrs[inter_frame_attr_nm] = attr
+
+                next_frame = t.tosys(**frattrs)
+
+                next_affine_params = t._affine_params(curr_frame, next_frame)
+                affine_params = _combine_affine_params(affine_params, next_affine_params)
+
+                # The frame needs to have data for the machinery, but that data should not be
```

Comment:
It's actually a little trickier than I thought because some existing affine transformations do look at the data instance because they check if the data has differentials before bothering to calculate and return the velocity shift.

Reply:
Added to the comment